### PR TITLE
SEP-0030: Add proposal for recoverysigner multi-party key management

### PR DESCRIPTION
**Pull request by leighmcculloch**
### What
Add a proposal for recoverysigner, multi-party key management of Stellar accounts where no server managing keys has control of the account, and supporting users:
1. Recovering access to Stellar accounts for which they may have lost keys.
2. Gaining access to Stellar accounts that another user intends to share with them.

### Why
I have been experimenting with this proposal in implementation first over in the [exp/services/recoverysigner] package which implements most of what is documented here. I think I've experimented enough with it to warrant documenting.

### Expectations
I expect this proposal to remain in draft state for sometime while the finer details of the SEP are ironed out. I'd like to merge the SEP without perfecting it so that I can iterate on it, and keep it and the experimental implementation in line so that the experimental implementation has a well defined specification as it evolves, and so that as the specification changes it is captured clearly outside of code.

### Reviewing in Style
To review the markdown rendered version, click here to view the file on the branch:
https://github.com/leighmcculloch/stellar--stellar-protocol/blob/sep-recovery-signer/ecosystem/sep-0030.md

[exp/services/recoverysigner]: https://github.com/stellar/go/blob/master/exp/services/recoverysigner

### CODE REVIEW

File: ecosystem/sep-0030.md [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+  "identities": {
+    "has_owner": true,
+    "has_other": true
+  },
+  "identity": "account",
+  "signer": "GADF..."
+}
+```
+
+#### `PUT /accounts/<address>`
```
Comment (accordeiro) [inline ecosystem/sep-0030.md]: :heart: the use of `PUT` here :)

File: ecosystem/sep-0030.md [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+- The server verifies the [JWT] is valid according to [SEP-10].
+- The server verifies that the account the [JWT] proves control of is the account being registered.
+- The server verifies that the account is not already registered.
+- The server stores the account address and alternative identities.
+- The server generates a unique random signing key for the account.
+- The server encrypts the signing secret key and stores it.
+- The server responds to the client with the signing public key.
+- The client adds the signing public key as a signer of the account with an appropriate weight.
+
+A client can register with multiple servers and give each signing key a weight that is appropriate for the use case the client and servers are targeting. As an example, if a client is intending to use two servers that should individually have no control of the account but together are able to approve transactions, then the client can configure [account thresholds] to high 20, med 20, low 20, and assign weights 10 to both server signing keys.
```
[the remark this answers is not in the document]
Comment (leighmcculloch) [reply ecosystem/sep-0030.md]: It is a link 😄. The actual URL is in the footnotes:
https://github.com/stellar/stellar-protocol/blob/c5ceb9a18b0fb1c1a7ed5545762f100dbba065fc/ecosystem/sep-0030.md#L479

File: ecosystem/sep-0030.md [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+- `SEP-10`: A [SEP-10] Web Auth server for proving high threshold control of an
+account. The [SEP-10] server should be configured to only issue [JWT]s if signers
+prove they meet the high threshold on the account.
+- `External`: Defined by the implementer. Proves control of a phone number,
+email address, or other identity.
+
+### Content Type
+
+All endpoints accept in requests the following `Content-Type`s:
+- `application/x-www-form-urlencoded`
```
[the remark this answers is not in the document]
Comment (leighmcculloch) [reply ecosystem/sep-0030.md]: We appear to offer the two content-types in other SEPs like SEP-10. @nebolsin introduced the concept and has shared some points for why it is beneficial, see stellar/stellar-protocol#150.

It's relatively easy for our reference implementations to offer both because [`"github.com/stellar/go/support/http/httpdecode".Decode`](https://godoc.org/github.com/stellar/go/support/http/httpdecode#Decode) functionality that unpacks both in a standard form. I'm not sure if it would be as easy in every programming language though and that's a tradeoff that might not be worth it.

We should maybe try and learn from SEP-10 and find out how many wallets and servers actually support url-encoded form bodies and leave it out if no one is using it.

I've created an issue to consider this: stellar/stellar-protocol#557. If we decide in that issue not to use it going forward I will remove it from this SEP if this SEP is still in draft state.

File: ecosystem/sep-0030.md [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+### Common Fields
+
+A set of common fields are used to describe an account in most endpoints:
+
+Name | Type | Description
+-----|------|------------
+`address` | string | The Stellar account ID or address of the account.
+`identities.has_owner` | boolean | True if the account has alternative identities set for the owner identity.
+`identities.has_other` | boolean | True if the account has alternative identities set for the other identity.
+`identities.*.account` | string | The Stellar account ID of an another account that if authenticated will be given full permissions of this account.
```
[the remark this answers is not in the document]
Comment (leighmcculloch) [reply ecosystem/sep-0030.md]: I was thinking of replacing the `*` with `owner|other`. It made the table wide though. I was also thinking that reading all these fields upfront is not necessarily very useful because the reader hasn't seen the context they are used in yet. What do you think if I move them to under the endpoints?

File: ecosystem/sep-0030.md [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+
+#### Status Code
+
+Common HTTP status codes may be returned for a server. In particular the following are expected:
+
+Status Code | Name | Reason
+-----|------|------------
+`400` | Bad Request | The request is invalid in anyway.
+`401` | Unauthorized | No `Authorization` header has been provided or the contents of the header are not accepted as valid.
+`404` | Not Found | The path is not recognized by the server, or the address does not have an account, or the authenticated user does not have permission to access the account.
```
[the remark this answers is not in the document]
Comment (leighmcculloch) [reply ecosystem/sep-0030.md]: Exactly as you say the choice of 404 is because the resource doesn't exist to somebody not authenticated or authenticated and doesn't have access.

Different APIs have interpreted this differently. AWS S3 uses 403 if you try and access anything you don't have access to, even if it actually doesn't exist. GitHub uses 404. I think 404 is more consistent with the actual descriptions of the status codes in our use case. It's also easier to implement because when implementing it as 403 it is really easy to accidentally return 404 in some cases and then leak insight.

It definitely makes the API a little more obtuse as it could be easy to think it doesn't exist if you aren't authenticated properly.

File: ecosystem/sep-0030.md [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+
+This protocol defines an API allowing a client to register an account with a
+server, and later ask the server to sign transactions for the account.
+
+The registration flow is as follows:
+- The client uses [SEP-10] to get a [JWT] proving high threshold control of an account.
+- The client, authenticated as the account, registers with the server, providing:
+  - Alternative identities for the owner of the account:
+    - A Stellar address.
+    - A phone number.
```
[the remark this answers is not in the document]
Comment (leighmcculloch) [reply ecosystem/sep-0030.md]: f7b725c

File: ecosystem/sep-0030.md [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+## Abstract
+
+This protocol defines an API allowing a client to register an account with a
+server, and later ask the server to sign transactions for the account.
+
+The registration flow is as follows:
+- The client uses [SEP-10] to get a [JWT] proving high threshold control of an account.
+- The client, authenticated as the account, registers with the server, providing:
+  - Alternative identities for the owner of the account:
+    - A Stellar address.
```
[the remark this answers is not in the document]
Comment (leighmcculloch) [reply ecosystem/sep-0030.md]: Ah yeah, good idea. I can't take credit for this idea, this was @tomquisel's input. 260e826

File: ecosystem/sep-0030.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (leighmcculloch) [reply ecosystem/sep-0030.md]: I've restructured this section, check it out. 904b04c

[the remark this answers is not in the document]
Comment (accordeiro) [reply ecosystem/sep-0030.md]: Looks much better, thanks!

File: ecosystem/sep-0030.md [hunk 8]
```diff
@@ -0,0 +1,479 @@
...
+Author: Leigh McCulloch <@leighmcculloch>, Lindsay Lin
+Track: Standard
+Status: Draft
+Created: 2019-12-20
+Version: 0.1.0
+```
+
+## Summary
+
+This protocol defines an API that servers can implement and a user, client, or wallet can use to regain access to a Stellar account that was registered with two or more implementing servers without any individual server controlling the account.
```
[the remark this answers is not in the document]
Comment (leighmcculloch) [reply ecosystem/sep-0030.md]: @accordeiro I took the content of your suggestion and shortened it into a single summary sentence. See 29a2786. Let me know if you think it's still too wordy. The template said to try and keep it simple, so I'm trying to make the summary not too elaborate.
